src/collectives/device/all_to_all.h

```diff
@@ -32,29 +43,55 @@ class ncclFunction<ncclFuncAllToAll, ALGO, PROTO, FUNC, T, UNROLL> {
+      const int workIndex = args->index+1;
+      volatile struct scklFlag* scklFlags = comm->scklFlags;
       // Compute pointers
-      const T * __restrict__ thisInput = (const T*)args->sendbuff;
-      T * __restrict__ thisOutput = (T*)args->recvbuff;
+      T * thisInput = (T*)args->sendbuff;
+      T * thisOutput = (T*)args->recvbuff;
       int myRank = channel->ring.devUserRanks[0];
       int m1 = -1;
       int recvPeer = (sckltb->type == SCKL_RECV) ? sckltb->peer : m1;
       int sendPeer = (sckltb->type == SCKL_SEND) ? sckltb->peer : m1;
 
       ncclPrimitives<UNROLL, ALLTOALL_CHUNKSTEPS/ALLTOALL_SLICESTEPS, ALLTOALL_SLICESTEPS, T, 1, 1, 1, FUNC>
         prims(tid, nthreads, &recvPeer, &sendPeer, thisOutput, stepSize, channel, comm, ncclShmem->ptrs, 0);
-
-      for (ssize_t gridOffset = 0; gridOffset < sizePerChunk; gridOffset += loopSize) {
+      for (ssize_t gridOffset = 0, iter = 0; gridOffset < sizePerChunk; gridOffset += loopSize, iter++) {
         int realChunkSize = min(chunkSize, DIVUP(sizePerChunk-gridOffset,nChannels));
         ALIGN_SIZE(realChunkSize, nthreads*sizeof(uint64_t)/sizeof(T));
         ssize_t chunkOffset = gridOffset + channelId*realChunkSize;
         ssize_t offset;
         int nelem = min(realChunkSize, sizePerChunk-chunkOffset);
         for (int i = 0; i < sckltb->nsteps; i++){
-          offset = chunkOffset + sckltb->transfers[i] * sizePerChunk;
+          struct scklTransfer* sckltran = &sckltb->transfers[i];
+          if (sckltran->offset == -1) continue;
+          offset = chunkOffset + sckltran->offset * sizePerChunk;
+          T* thisbuffer = (sckltran->buffer == SCKL_THIS_INPUT) ? thisInput : thisOutput;
           if (sckltb->type == SCKL_SEND){
-            prims.directSend(thisInput + offset, offset, nelem);
+            int8_t dependentBid = sckltran->dependentRbid + scklNumBlocksPerChannel * channelId;
+            int8_t dependentStep = sckltran->dependentStep;
+            if (sckltran->dependentRbid >= 0){
+              if (tid == 0){
+                uint64_t readFlag;
+                int readGlobalIterStep;
+                int readWorkIndex;
+                int gaolGlobalIterStep = GLOBALITERSTEP(iter, dependentStep);
+                do {
+                  readFlag = (scklFlags + dependentBid)->flag;
+                  GETFLAGITEMS(readGlobalIterStep, readWorkIndex, readFlag);
+                } while (readWorkIndex != workIndex || readGlobalIterStep < gaolGlobalIterStep);
+              }
+              __syncthreads();
+            }
+            prims.directSend(thisbuffer + offset, offset, nelem);
           } else if (sckltb->type == SCKL_RECV) {
-            prims.directRecv(thisOutput + offset, offset, nelem);
+            prims.directRecv(thisbuffer + offset, offset, nelem);
+            if (tid == 0){
+              uint64_t curFlag;
+              SETFLAG(iter, i, workIndex, curFlag);
```

Comment:
make SETFLAG a "function" like COMPUTE_FLAG and then replace this with curFlag = COMPUTE_FLAG(...)

Reply:
Done!